File: java/twelve-days/src/main/java/TwelveDays.java
class TwelveDays {
    private static final String[] ORDINALS = {
        "first", "second", "third", "fourth", "fifth", "sixth",
        "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"
    };
    
    private static final String[] GIFTS = {
        "twelve Drummers Drumming",
        "eleven Pipers Piping",
        "ten Lords-a-Leaping",
        "nine Ladies Dancing",
        "eight Maids-a-Milking",
        "seven Swans-a-Swimming",
        "six Geese-a-Laying",
        "five Gold Rings",
        "four Calling Birds",
        "three French Hens",
        "two Turtle Doves",
        "a Partridge in a Pear Tree"
    };
    
    String verse(int verseNumber) {
        StringBuilder verse = new StringBuilder();
        
        // Add the beginning of the verse
        verse.append(String.format("On the %s day of Christmas my true love gave to me: ", 
                ORDINALS[verseNumber - 1]));
        
        // Add the gifts
        for (int i = 12 - verseNumber; i < 11; i++) {
            verse.append(GIFTS[i]).append(", ");
        }
        
        // Add the last gift (with "and" if not the first verse)
        if (verseNumber > 1) {
            verse.append("and ");
        }
        verse.append(GIFTS[11]).append(".\n");
        
        return verse.toString();
    }
    
    String verses(int startVerse, int endVerse) {
        StringBuilder verses = new StringBuilder();
        
        for (int i = startVerse; i <= endVerse; i++) {
            verses.append(verse(i));
            
            // Add a blank line between verses, but not after the last verse
            if (i < endVerse) {
                verses.append("\n");
            }
        }
        
        return verses.toString();
    }
    
    String sing() {
        return verses(1, 12);
    }
}

File: javascript/twelve-days/twelve-days.js
export const recite = (start, end = start) => {
  const ordinals = [
    'first',
    'second',
    'third',
    'fourth',
    'fifth',
    'sixth',
    'seventh',
    'eighth',
    'ninth',
    'tenth',
    'eleventh',
    'twelfth'
  ];

  const gifts = [
    'a Partridge in a Pear Tree',
    'two Turtle Doves',
    'three French Hens',
    'four Calling Birds',
    'five Gold Rings',
    'six Geese-a-Laying',
    'seven Swans-a-Swimming',
    'eight Maids-a-Milking',
    'nine Ladies Dancing',
    'ten Lords-a-Leaping',
    'eleven Pipers Piping',
    'twelve Drummers Drumming'
  ];

  const generateVerse = (day) => {
    const dayIndex = day - 1;
    let verse = `On the ${ordinals[dayIndex]} day of Christmas my true love gave to me: `;
    
    if (day === 1) {
      verse += `${gifts[0]}.`;
    } else {
      for (let i = dayIndex; i > 0; i--) {
        verse += `${gifts[i]}, `;
      }
      verse += `and ${gifts[0]}.`;
    }
    
    return verse + '\n';
  };

  let result = '';
  for (let day = start; day <= end; day++) {
    result += generateVerse(day);
    if (day < end) {
      result += '\n';
    }
  }
  
  return result;
};
